From fde257ab074aee66204e9220c4d59595b2acff8d Mon Sep 17 00:00:00 2001
From: Eneas U de Queiroz <cotequeiroz@gmail.com>
Date: Thu, 21 Nov 2019 14:28:23 -0300
Subject: [PATCH] crypto: qce - fix xts-mode key sizes

XTS-mode uses two keys, so the keysizes in the driver definition should
be doubled, and halved when checking if it is AES-128/192/256.

Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

diff --git a/drivers/crypto/qce/skcipher.c b/drivers/crypto/qce/skcipher.c
index a22329a6c4cc..b26340c9b531 100644
--- a/drivers/crypto/qce/skcipher.c
+++ b/drivers/crypto/qce/skcipher.c
@@ -247,7 +247,7 @@ static int qce_skcipher_setkey(struct crypto_skcipher *ablk, const u8 *key,
 	}
 
 	if (IS_AES(flags)) {
-		switch (keylen) {
+		switch (IS_XTS(flags) ? keylen >> 1 : keylen) {
 		case AES_KEYSIZE_128:
 		case AES_KEYSIZE_256:
 			break;
@@ -393,8 +393,8 @@ static const struct qce_skcipher_def skcipher_def[] = {
 		.drv_name	= "xts-aes-qce",
 		.blocksize	= AES_BLOCK_SIZE,
 		.ivsize		= AES_BLOCK_SIZE,
-		.min_keysize	= AES_MIN_KEY_SIZE,
-		.max_keysize	= AES_MAX_KEY_SIZE,
+		.min_keysize	= AES_MIN_KEY_SIZE * 2,
+		.max_keysize	= AES_MAX_KEY_SIZE * 2,
 	},
 	{
 		.flags		= QCE_ALG_DES | QCE_MODE_ECB,
